core/services/keeper/registry_synchronizer.go

```diff
@@ -0,0 +1,221 @@
+	keeperORM KeeperORM,
+	syncInterval time.Duration,
+) RegistrySynchronizer {
+	return RegistrySynchronizer{
+		contract:  contract,
+		keeperORM: keeperORM,
+		job:       job,
+		interval:  syncInterval,
+		isRunning: atomic.NewBool(false),
+		chDone:    make(chan struct{}),
+	}
+}
+
+// RegistrySynchronizer conforms to the job.Service interface
+var _ job.Service = RegistrySynchronizer{}
+
+type RegistrySynchronizer struct {
+	contract  *keeper_registry_wrapper.KeeperRegistry
+	interval  time.Duration
+	isRunning *atomic.Bool
+	job       job.SpecDB
+	keeperORM KeeperORM
+
+	chDone chan struct{}
+}
+
+func (rs RegistrySynchronizer) Start() error {
+	if rs.isRunning.Load() {
+		return errors.New("already started")
+	}
+	rs.isRunning.Store(true)
+	go rs.run()
+	return nil
+}
+
+func (rs RegistrySynchronizer) Close() error {
+	close(rs.chDone)
+	return nil
+}
+
+func (rs RegistrySynchronizer) run() {
+	ticker := time.NewTicker(rs.interval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-rs.chDone:
+			return
+		case <-ticker.C:
+			rs.syncRegistry()
+		}
+	}
+}
+
+func (rs RegistrySynchronizer) syncRegistry() {
+	contractAddress := rs.job.KeeperSpec.ContractAddress
+	logger.Debugf("syncing registry %s", contractAddress.Hex())
+
+	err := func() error {
+		registry, err := rs.newSyncedRegistry(rs.job)
+		if err != nil {
+			return err
+		}
+		if err = rs.keeperORM.UpsertRegistry(&registry); err != nil {
+			return err
+		}
+		if err = rs.addNewUpkeeps(registry); err != nil {
+			return err
+		}
+		if err = rs.deleteCanceledUpkeeps(registry); err != nil {
+			return err
+		}
+		return nil
+	}()
+
+	if err != nil {
+		logger.Errorf("unable to sync registry %s, err: %v", contractAddress.Hex(), err)
+	}
+}
+
+func (rs RegistrySynchronizer) addNewUpkeeps(reg Registry) error {
+	nextUpkeepID, err := rs.keeperORM.NextUpkeepIDForRegistry(reg)
+	if err != nil {
+		return err
+	}
+
+	countOnContractBig, err := rs.contract.GetUpkeepCount(nil)
+	if err != nil {
+		return err
+	}
+	countOnContract := countOnContractBig.Int64()
+
+	wg := sync.WaitGroup{}
+	wg.Add(int(countOnContract - nextUpkeepID))
+
+	// batch sync registries
+	chSyncUpkeepQueue := make(chan struct{}, syncUpkeepQueueSize)
+	done := func() { <-chSyncUpkeepQueue; wg.Done() }
+	for upkeepID := nextUpkeepID; upkeepID < countOnContract; upkeepID++ {
+		chSyncUpkeepQueue <- struct{}{}
+		go rs.syncUpkeep(reg, upkeepID, done)
+	}
+
+	wg.Wait()
+	return nil
+}
+
+func (rs RegistrySynchronizer) deleteCanceledUpkeeps(reg Registry) error {
+	canceledBigs, err := rs.contract.GetCanceledUpkeepList(nil)
+	if err != nil {
+		return err
+	}
+	canceled := make([]int64, len(canceledBigs))
+	for idx, upkeepID := range canceledBigs {
+		canceled[idx] = upkeepID.Int64()
+	}
+	return rs.keeperORM.BatchDeleteUpkeeps(reg.ID, canceled)
+}
+
+func (rs RegistrySynchronizer) syncUpkeep(registry Registry, upkeepID int64, doneCallback func()) {
+	defer doneCallback()
+
+	err := func() error {
```

Comment:
it's just so that the error logging can be combined into one step. At the end of `syncUpkeep` is a logger statement that logs the error. If I remove this anonymous function, then we have to add 3 logger statements to handle each error case. I think this keeps things simpler.

Reply:
I think a 
```
var err error
defer func() {
         // log err
}()
```
would be simpler / more readable.